src/dataflow/src/source/s3.rs

```diff
@@ -0,0 +1,451 @@
+                    e
+                );
+                Err(e)
+            }
+            Err(TryRecvError::Empty) => Ok(NextMessage::Pending),
+            Err(TryRecvError::Disconnected) => Ok(NextMessage::Finished),
+        }
+    }
+
+    fn can_close_timestamp(
+        &self,
+        consistency_info: &ConsistencyInfo,
+        pid: &PartitionId,
+        offset: MzOffset,
+    ) -> bool {
+        if !self.is_activated_reader {
+            true
+        } else {
+            // TODO: when is this ever not true for S3?
+            let last_offset = consistency_info
+                .partition_metadata
+                .get(&pid)
+                // Guaranteed to exist for S3
+                .unwrap()
+                .offset;
+            last_offset >= offset
+        }
+    }
+
+    fn get_worker_partition_count(&self) -> i32 {
+        panic!("s3 sources do not support BYO consistency: get_worker_partition_count")
+    }
+
+    fn has_partition(&self, _partition_id: PartitionId) -> bool {
+        panic!("s3 sources do not support BYO consistency: has_partition")
+    }
+
+    fn ensure_has_partition(&mut self, _consistency_info: &mut ConsistencyInfo, _pid: PartitionId) {
+        panic!("s3 sources do not support BYO consistency: ensure_has_partition")
+    }
+
+    fn update_partition_count(
+        &mut self,
+        consistency_info: &mut ConsistencyInfo,
+        partition_count: i32,
+    ) {
+        log::debug!(
+            "ignoring partition count update type={:?} partition_count={}",
+            consistency_info.source_type,
+            partition_count,
+        )
+    }
+
+    fn buffer_message(&mut self, message: SourceMessage<Out>) {
+        if self.buffer.is_some() {
+            log::error!("Internal error: buffer is not empty, overwriting with new message");
+        }
+        self.buffer = Some(message);
+    }
+}
+
+// Helper utilities
+
+/// Find the unambiguous prefix of a glob
+fn find_prefix(glob: &str) -> String {
+    let mut escaped = false;
+    let mut escaped_filter = false;
+    glob.chars()
+        .take_while(|c| match (c, &escaped) {
+            ('*', false) => false,
+            ('[', false) => false, // a character class is a form of glob
+            ('{', false) => false, // a group class is a form of glob
+            ('\\', false) => {
+                escaped = true;
+                true
+            }
+            (_, false) => true,
+            (_, true) => {
+                escaped = false;
+                true
+            }
+        })
+        .filter(|c| match (c, &escaped_filter) {
+            (_, true) => {
+                escaped_filter = false;
+                true
+            }
+            ('\\', false) => {
+                escaped_filter = true;
+                false
+            }
+            (_, _) => true,
+        })
+        .collect()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn glob_prefix() {
```

Comment:
What happens when `\` appears at the end of the string? Should we test that case?

Reply:
That's impossible because globset refuses to create a glob with a dangling escape:

```rust
        println!(
            "{}",
            globset::GlobBuilder::new(r"f\")
                .literal_separator(true)
                .backslash_escape(true)
                .build()
                .unwrap_err()
        )
```
prints: `error parsing glob 'f\': dangling '\'`